doc/contributing/erm-guidelines.md

```diff
@@ -0,0 +1,329 @@
+be used with the `await using` statement in an asynchronous context. It is
+worth noting here that `await using` means the disposal is asynchronous,
+not the initialization.
+
+```mjs
+{
+  using resource = new MyResource();
+  await using asyncResource = new MyResource();
+}
+```
+
+Importantly, it is necessary to understand that the design of ERM makes it
+possible for user code to call the `Symbol.dispose` or `Symbol.asyncDispose`
+methods directly, outside of the `using` or `await using` statements. These
+can also be called multiple times and by any code that is holding a reference
+to the object. That is to say, ERM does not imply ownership of the object. It
+is not a form of RAII (Resource Acquisition Is Initialization) as seen in some
+other languages and there is no notion of exclusive ownership of the object.
+A disposable object can become disposed at any time.
+
+The `Symbol.dispose` and `Symbol.asyncDispose` methods are called in both
+successful and exceptional exits from the scopes in which the using keyword
+is used. This means that if an exception is thrown within the scope, the
+disposal methods will still be called. However, when the disposal methods are
+called they are not aware of the context in which they were called. These
+methods will not receive any information about the exception that was thrown
+or whether an exception was thrown at all. This means that it is safest to
+assume that the disposal methods will be called in a context where the object
+may not be in a valid state or that an exception may be pending.
+
+## Guidelines for Disposable Objects
+
+So with this is mind, it is necessary to outline some guidelines for disposers:
+
+1. Disposers should be idempotent. Multiple calls to the disposal methods
+   should not cause any issues or have any additional side effects.
+2. Disposers should assume that it is being called in an exception context.
+   Always assume there is likely a pending exception and that if the object
+   has not been explicitly closed when the disposal method is called, that
+   the object should be disposed as if an exception had occurred. For instance,
+   if the object API exposes both a `close()` method and an `abort()` method,
+   the disposal method should call `abort()` if the object is not already
+   closed. If there is no difference in disposing in success or exception
+   contexts, then separate disposal methods are unnecessary.
+3. Disposers may throw their own exceptions but this is not recommended.
+   If a disposer throws an exception while there is another pending
+   exception, then both exceptions will be wrapped in a `SupressedError`
+   that masks both. This makes it difficult to understand the context
+   in which the exceptions were thrown. It is, however, not possible to
+   completely prevent exceptions from being thrown in the disposal methods
+   so this guideline is more of a recommendation than a hard rule.
```

Comment:
```suggestion
3. It is recommended to avoid throwing errors within disposers.
   If a disposer throws an exception while there is another pending
   exception, then both exceptions will be wrapped in a `SupressedError`
   that masks both. This makes it difficult to understand the context
   in which the exceptions were thrown.
```